packages/xo-server-load-balancer/src/plan.js

```diff
@@ -777,11 +809,311 @@ export default class Plan {
+          this._thresholds.memoryFree.critical
+        ) {
+          debugAffinity(`Host ${sourceHost.id} is overcrowded`)
+          // A) migrate other VMs to try to free some memory on destination host
+          const { promises: otherMigrationPromises, success } = await this._migrateOtherVms({
+            crowdedHost: destinationHost,
+            hostsAverages,
+            vmsAverages,
+            idToHost,
+            taggedHosts,
+            memoryNeeded: vmsAverages[vm.id].memory,
+          })
+          promises.push(...otherMigrationPromises)
+
+          // B) if we can't do A), change the destination to the next host to create another host with several VMs with that tag
+          if (!success) {
+            debugAffinity(
+              `Host ${sourceHost.id} does not have enough memory to get all "${tag}" tagged VMs (or its coalition)`
+            )
+            if (sourceHost === sortedHosts[sortedHosts.length - 1]) {
+              warn(`affinity: Can't satisfy ${tag} affinity constraints (or its coalition)`)
+              return promises
+            }
+            destinationHost = sortedHosts.pop()
+          }
+        }
+
+        promises.push(
+          this._migrateVmAndUpdateInfos({
+            destination: idToHost[destinationHost.id],
+            source: idToHost[sourceHost.id],
+            sourceHost,
+            destinationHost,
+            vm,
+            hostsAverages,
+            vmAverages: vmsAverages[vm.id],
+          })
+        )
+      }
+    }
+    return promises
+  }
+
+  async _migrateOtherVms({ crowdedHost, hostsAverages, vmsAverages, idToHost, taggedHosts, memoryNeeded }) {
+    const promises = []
+
+    const candidateVms = sortBy(
+      filter(
+        Object.values(crowdedHost.vms),
+        vm => vm.xenTools && intersection(vm.tags, this._affinityTags).length === 0
+      ),
+      [vm => -vmsAverages[vm.id].memory] // try to migrate bigger VMs first to minimize the number of migrations
+    )
+    debugAffinity(`Candidate VMs to be moved away: ${candidateVms.map(vm => vm.name_label)}`)
+
+    for (const vm of candidateVms) {
+      // try to migrate vm
+
+      const vmAverages = vmsAverages[vm.id]
+
+      const destinationId = sortBy(
+        Object.keys(hostsAverages),
+        [hostId => -hostsAverages[hostId].memoryFree, hostId => hostsAverages[hostId].cpu] // try to migrate to hosts with the most free space first
+      ).find(hostId => {
+        if (hostId === crowdedHost.id) {
+          return false
+        }
+        const destinationAverages = hostsAverages[hostId]
+        return (
+          destinationAverages.cpu + vmAverages.cpu <= this._thresholds.cpu.critical &&
+          destinationAverages.memoryFree - vmAverages.memory >= this._thresholds.memoryFree.critical
+        )
+      })
+
+      if (destinationId === undefined) {
+        // no host can accept this VM, let's try another one
+        debug(`Cannot migrate VM (${vm.id}) to any Host.`)
```

Comment:
```suggestion
        debug(`Cannot migrate VM (${vm.id}) to any Host. VM requires ${vmAverages.memory}MB, CPU: ${vmAverages.cpu}%`)
```
In my opinion, this information could be interesting to have.